### ISSUE TITLE
Don't restrict partition sizes to exactly fit on the right or bottom frame boundaries

### ISSUE THREAD

**Issue by ycho** [xiph/rav1e#2166]
* Current block partitioning approach on right or borrtom frame boundaries in rav1e
- Both topdown and bottomup function keep splitting the input SB (SuperBlock) until no partitioned blocks straddle on the right or bottom frame boundaries.
(The condition whether keep splitting or not is checked by 'must_split', which is set true if the current partition size straddle on the frame boundary or other condition such as if current size is larger than desired max partition size as shown in the code, https://github.com/xiph/rav1e/blob/8f273bcbde77e5f3711138a57c13dec9dc793973/src/encoder.rs#L2546)

Affected areas of codebase can be:
- [x] Partition search functions should not restrict partition sizes to exactly fit on the right or bottom frame boundaries
- [x] Distortion compute functions to use visible area, which can be any size not defined by av1 partition sizes
~~- [x] Intra prediction to predict only visible pixels only~~
- [x] Ref pixels for intra pred (i.e. intra edge pixels) should not use invisible pixels.
~~- [ ] Inter prediction (Motion Estimation) to predict only visible pixels only~~
~~- [ ] Define the pixel values for invisible area as input to forward transforms, i.e. what kind of padding to use for input and reconstructed frame? Already defined that extension of directly adjacent and last available pixel.~~

**Comment by ycho** [comment]
Example of partitions decided by current rav1e. 
The 2nd frame (inter frame) of 'blue_sky_360p_60f.y4m' encoded by rav1e, -s 6:
There are 10x6 (row x col) SBs (SuperBlocks) are encoded for a frame.

The bottom SB row has 40 visible pixels in height.
See that all SBs in the bottom-most SB row are SPLIT down to: two of 32x32 and two of 16x4 partitions. Our goal is to enable rav1e to not split for this SB, which is allowed by av1 spec.
Note that av1 only allows PARTITION_SPLIT_NONE if more than half of block size pixels available in both row and col, as stated in https://aomediacodec.github.io/av1-spec/#decode-partition-syntax.
<img width="1678" alt="Screen Shot 2020-02-13 at 9 27 12 AM" src="https://user-images.githubusercontent.com/10123259/74461249-0f9dda00-4e43-11ea-8a7d-998e314d9bf2.png">

Captured from https://beta.arewecompressedyet.com/analyzer/?maxFrames=4&decoder=https://people.xiph.org/~mbebenita/analyzer/inspect.js&decoderName=master-1955d6d49b692e687661ba3bed95a26e91877645&file=https://beta.arewecompressedyet.com/runs/master-1955d6d49b692e687661ba3bed95a26e91877645/objective-1-fast/blue_sky_360p_60f.y4m-252.ivf

**Comment by ycho** [comment]
On the contrary, here is the partitions decided by libaom (+ my modification that min partition size = 64x64)

<img width="1653" alt="Screen Shot 2020-02-13 at 9 47 40 AM" src="https://user-images.githubusercontent.com/10123259/74463024-006c5b80-4e46-11ea-9fe6-c046e3ff7e93.png">

And its motion vectors:
<img width="1161" alt="Screen Shot 2020-02-13 at 9 49 02 AM" src="https://user-images.githubusercontent.com/10123259/74463075-167a1c00-4e46-11ea-8ea9-37140f2cf3f9.png">

**Comment by tmatth** [comment]
@ycho nice analysis, do you know if/how libaom is behaving compared to rav1e for this?

**Comment by ycho** [comment]
> ~@ycho nice analysis, do you know if/how libaom is behaving compared to rav1e for this?~ Nevermind, you just answered this before I asked 👍

Hey~ thanks for trying asking and enjoying my art of screen-captueing!

**Comment by ycho** [comment]
@barrbrain, if bsize straddle on frame borders (with my dev branch), luma_ac() gets panic at https://github.com/xiph/rav1e/blob/8f273bcbde77e5f3711138a57c13dec9dc793973/src/encoder.rs#L1859. I think it is because the subregion() in plane_regions.rs limit the w and h of region only up to tile (or frame if one tile) w and h, https://github.com/xiph/rav1e/blob/8f273bcbde77e5f3711138a57c13dec9dc793973/src/encoder.rs#L1852.

**Comment by ycho** [comment]
Little progress with enforce partitions in a bottom SB row all skipped, where revised partition decide to choose 64x64 because split into four 32x32s is not available since two 32x32 of them requires to be further partitioned by the definition of av1 spec, which topdown partition does not do.

<img width="1365" alt="Screen Shot 2020-02-24 at 4 14 22 PM" src="https://user-images.githubusercontent.com/10123259/75202759-2684dc00-5721-11ea-96d8-0319cde4134f.png">

Speed 9 is used, where 32x32 or 64x64 partition sizes are available.
See the bottom SB row where 64x64 are able to be used now, which does not exactly matches the frame size.

**Comment by ycho** [comment]
When computing the residue signal for tx input, rav1e has no definition what to do about outside frame pixels, which is to set outside residue values as zeros. So, for the tx-block that straddle on frame border, while intra prediction can be done for whole tx-block, the residue block for it should not count outside frame pixels.
Below decoded image is by dav1d, since aomdec failed to decode it (hence no image from AOMAnalyzer)
Hopefully unrelated to my changes, we have some new issues occurring on the last SB column for unknown reason. Note that full SB is inside the frame.

![Screen Shot 2020-03-03 at 11 27 02 AM](https://user-images.githubusercontent.com/10123259/75812003-4e3bfb80-5d42-11ea-8c34-e864c2b3016b.png)

**Comment by ycho** [comment]
So, "let mut residual_storage: Aligned<[i16; 64 * 64]> = Aligned::uninitialized();" in encode_tx_block(), can fill the array with values > abs(255), I think any value of i16. If I init with  any random values in that range, it works with my test (fill the out of tile residue values, which is input to fwd tx).
Then the best values to fill out of tile residue will be those that can generate lesser bits, for ex smaller magnitude and less AC coeffs. 
Also then, the convention that encoder does prediction for whole block make some sense if original source frame is reasonably padded on bottom/right frame aprons with the values that are close enough to predicted values.

**Comment by ycho** [comment]
Fixed the corrupted last SB column.
- When preparing ref pixels for intra prediction, i.e, inside
    get_intra_edges(), and for two cases, need_left and need_top,
    tx_size is clipped incorrectly when tx block straddle on the right or
    bottom frame border.

![Screen Shot 2020-03-05 at 4 50 31 PM](https://user-images.githubusercontent.com/10123259/76039734-25fafb00-5f02-11ea-8653-446fef4b3760.png)

**Comment by ycho** [comment]
Hmm, I don't know what inverse_transform_add() does when tx-block straddle on frame border.

**Comment by ycho** [comment]
To investigate the corrupted pixels on open partition, I am testing function by function in major path of encoding and reconstructing a tx-block, i.e. encode_tx_block(), one of them is wrting diff() (generating residue signal between original source and predicted pixels in a block) as below.

```rust
fn diff2<T: Pixel>(
  dst: &mut [i16], src1: &PlaneRegion<'_, T>, src2: &PlaneRegion<'_, T>,
  width: usize, height: usize,
) {
  let stride1 = src1.plane_cfg.stride;
  let stride2 = src2.plane_cfg.stride;

  for y in 0..height {
    for x in 0..width {
      unsafe {
      let v1 = src1.data_ptr().add(y * stride1 + x);
      let v2 = src2.data_ptr().add(y * stride2 + x);
      dst[y * width + x] = i16::cast_from(*v1) - i16::cast_from(*v2);
      }
    }
  }
}

**Comment by ycho** [comment]
I've tried this experiment of open partition on older commit, Nov 2018, where non-square partition in not introduced yet.
https://github.com/ycho/rav1e/commits/708a806c_2018_1104

With first commit, try encode at speed 10, then can see bottom 64x64 partitions are all corrupted.
(ex: ./target/release/rav1e nyan.y4m -o test.ivf -r test_rec.y4m --quantizer 50 --speed=10 --limit=2)
https://github.com/ycho/rav1e/commit/0cee6bef8f5c615aaf7ea5bc37f9f6da8d902372

With 2nd commit, which skips on bottom 64x64 partitions (if they are on frame borders), then decoder (aomdec) can decode the bitstream.

This makes me think somewhere in coefficient (i.e. residue pixels) encoding stage is wrong for open partition.

**Comment by ycho** [comment]
48x48 size test input.
[twitter.y4m.gz](https://github.com/xiph/rav1e/files/4423080/twitter.y4m.gz)

**Comment by ycho** [comment]
48x48 size test input.
[twitter.y4m.gz](https://github.com/xiph/rav1e/files/4423080/twitter.y4m.gz)

**Comment by ycho** [comment]
New test images to test with 2x2, 3x3, 3x2 SB cases.

https://ibb.co/5jbhwCG
https://ibb.co/P1stvTL
https://ibb.co/W26mZqQ
https://ibb.co/VttWLFt

**Comment by ycho** [comment]
Currently, I can reproduce corruption of u and v channel for 128x112 input, https://github.com/xiph/rav1e/files/4458167/test03_128x112.y4m.gz
with the work branch
https://github.com/ycho/rav1e/commit/b21c744437a9022a6312bf09d1b968c82e04ad39
and enc command:
./target/release/rav1e test03_128x112.y4m -o test.ivf -r test_rec.y4m --quantizer 50 --speed=9 --limit=1 --rdo-lookahead-frames=1 --low_latency

**Comment by ycho** [comment]
Currently, it works with 112x112 (i.e. wxh = 1.75 x 1.75 SBs), but does not work with 128x112 (2x1.75 SBs)
<img width="846" alt="Screen Shot 2020-04-14 at 4 32 19 PM" src="https://user-images.githubusercontent.com/10123259/79388146-b07f4300-7f21-11ea-8d3f-3c2128e2c969.png">

**Comment by ycho** [comment]
I've found the reason finally!
<img width="1006" alt="Screen Shot 2020-04-20 at 8 05 38 PM" src="https://user-images.githubusercontent.com/10123259/79940173-cefba780-8415-11ea-8a47-90f7317b5f6e.png">

**Comment by ycho** [comment]
Brand new  test input to test 4x4 partitions sizes (and all other sizes as well :) ).
[test03_254x254.y4m.gz](https://github.com/xiph/rav1e/files/4525462/test03_254x254.y4m.gz)

![test03_254x254](https://user-images.githubusercontent.com/10123259/80152550-b9e75b80-8570-11ea-8648-4c37dde5f71b.png)

**Comment by ycho** [comment]
One example from objective_1_fast test sequence set, where open partition happens vs current master uses partitions to match frame boundaries.

<img width="1382" alt="Screen Shot 2020-05-08 at 3 23 26 PM" src="https://user-images.githubusercontent.com/10123259/81596059-48811880-9378-11ea-99cf-b3daf9fdf44a.png">

**Comment by ycho** [comment]
In inter frame,

<img width="1381" alt="Screen Shot 2020-05-08 at 3 25 46 PM" src="https://user-images.githubusercontent.com/10123259/81596180-79614d80-9378-11ea-8a00-c9853d523f38.png">

**Comment by ycho** [comment]
Topdown partition with bottom-up partition of frame boundary:
At speed 6, I still have ~1.5% coding loss.

https://beta.arewecompressedyet.com/?job=test_open_partition_btmup_on_boundary_s6%402020-05-27T23%3A16%3A14.788Z&job=master_af7f474fb%402020-05-25T16%3A43%3A25.801Z

For one of largest regression, I captured some screen shots for "rush_hour_1080p25_60f.y4m", at QIndex 172 for 1st frame.

master:
<img width="1749" alt="Screen Shot 2020-06-03 at 10 32 32 AM" src="https://user-images.githubusercontent.com/10123259/83670720-74af4400-a588-11ea-8b44-5ad72b27f732.png">

open-partition branch
<img width="1749" alt="Screen Shot 2020-06-03 at 10 32 41 AM" src="https://user-images.githubusercontent.com/10123259/83670735-7aa52500-a588-11ea-9120-6c7cdb5ac8db.png">

Then, partition info.
master
<img width="1749" alt="Screen Shot 2020-06-03 at 10 32 51 AM" src="https://user-images.githubusercontent.com/10123259/83670774-855fba00-a588-11ea-804f-e34992406e4a.png">

open-partition branch
<img width="1749" alt="Screen Shot 2020-06-03 at 10 32 58 AM" src="https://user-images.githubusercontent.com/10123259/83670786-8b559b00-a588-11ea-80b2-d35a1b85bedc.png">
Please find and see the "Bits" (per frame) info on the middle of right panel, that shows 130,982 - 130,477 = 505 bits less than master.

**Comment by ycho** [comment]
Result at speed 2

https://beta.arewecompressedyet.com/?job=test_open_partition_btmup_on_boundary_s2%402020-05-27T23%3A17%3A13.047Z&job=master_c6bf0bf_s2%402020-05-20T23%3A40%3A23.097Z

**Comment by ycho** [comment]
Result at speed 1

https://beta.arewecompressedyet.com/?job=master_af7f474fb_s1%402020-05-27T21%3A09%3A32.074Z&job=test_open_partition_fix2_s1%402020-05-27T06%3A14%3A22.687Z

**Comment by ycho** [comment]
Investigations on open partition scheme why it introduces coding loss for both bottom-up and top-down, ~1.5%.  Visually checking with bitstream analyzer, open partitions seem reasonably chosen during partition search, yielding less bits but at larger QPs the quality seems worse compared to reference master.
Experiment using bottom-up partition for top-down on the frame boundary, otherwise, top-down just stops with larger open partition.
And, this leads me to think over the possibility of lambda used by rav1e might not be right for frame boundary case since it never has been trained from such cases. I.e seemingly too harsh penalization for partition cost for SB on frame boundary. But this is a hypothesis, just my idea needing a test (and hopefully proof), probably I need to start from knowing the origin of rav1e’s lambda and how it has formed.

**Comment by ycho** [comment]
FYI, current work branch is https://github.com/ycho/rav1e/commits/test_open_partition_use_bottomup_on_frame_boundary

**Comment by ycho** [comment]
For QIndex = 253 (but the QIndex for 1st frame is 231),

https://beta.arewecompressedyet.com/runs/master_af7f474fb@2020-05-25T16:43:25.801Z/objective-1-fast/rush_hour_1080p25_60f.y4m-252.ivf
https://beta.arewecompressedyet.com/runs/test_open_partition_btmup_on_boundary_s6@2020-05-27T23:16:14.788Z/objective-1-fast/rush_hour_1080p25_60f.y4m-252.ivf

With SB grid .
master
<img width="1716" alt="Screen Shot 2020-06-07 at 4 59 20 PM" src="https://user-images.githubusercontent.com/10123259/83983122-b7528280-a8e0-11ea-9f21-0886feca58b3.png">

open-partion
<img width="1716" alt="Screen Shot 2020-06-07 at 4 59 24 PM" src="https://user-images.githubusercontent.com/10123259/83983124-c0435400-a8e0-11ea-930b-52cb7081c133.png">

With partition info
master
<img width="1716" alt="Screen Shot 2020-06-07 at 4 59 31 PM" src="https://user-images.githubusercontent.com/10123259/83983127-c5a09e80-a8e0-11ea-8536-e72801c71b6f.png">

open-partion
<img width="1716" alt="Screen Shot 2020-06-07 at 4 59 37 PM" src="https://user-images.githubusercontent.com/10123259/83983132-cb967f80-a8e0-11ea-8152-1ac70f9d19fa.png">

**Comment by ycho** [comment]
Result for still image sets, subset1
For [awcy mono subset1, speed 0](https://beta.arewecompressedyet.com/?job=test_open_partition_fix4_1f_s0_mono%402020-06-14T07%3A53%3A58.849Z&job=master_af7f474fb_1f_s0_mono%402020-06-11T20%3A42%3A31.682Z)

[awcy subset1, speed 0](https://beta.arewecompressedyet.com/?job=master_af7f474fb_1f_s0subset1%402020-06-14T15%3A56%3A35.870Z&job=test_open_partition_fix4_cfl_off_1f_s0_subset1%402020-06-14T08%3A18%3A40.954Z)

There were several fixes for varying image frame sizes that are not found in objective-1-fast.
https://github.com/ycho/rav1e/commit/805d68d9f521158c311b94185067b653bb99c3b5
https://github.com/ycho/rav1e/commit/aaa36d0b5bc189b32b51a5db0b04670dc9976023
https://github.com/ycho/rav1e/commit/3ead6b6feaabfc02885a15dae58a7187dec59b0d

They mostly address the fix of size of tx contexts in w and h for coefficient encoder, which occurs with two opposite cases, 1) when tx blocks are not visible but required to be encoded, for ex sub 8x8 blocks not inside visible area but inside the encoded frame, 2) part of tx block is not visible but inside the encoded frame.

**Comment by ycho** [comment]
The work branch has moved to https://github.com/ycho/rav1e/tree/make_partition_strong, and I plan to split one large patch into many smaller ones. Thank you!

**Comment by ycho** [comment]
Adding more test images:

[test03_252x252.y4m.gz](https://github.com/xiph/rav1e/files/4794458/test03_252x252.y4m.gz)
[test03_253x253.y4m.gz](https://github.com/xiph/rav1e/files/4794459/test03_253x253.y4m.gz)

**Comment by ycho** [comment]
> Another possibility for the coding loss by open partition is the undefined satd when pre-screening intra modes. It is good that satd compute with unsafe { } (since it can cross the plane region), but I had no chance what it does on frame boundaries. So, one way to exclude this factor is not doing intra mode decision then fix all intra mode block uses one mode, like dc pred.
> I will try that soon.

Checked this, and it seems not the reason from seeing [awcy result for 1 frame only at speed 1](https://beta.arewecompressedyet.com/?job=test_master_dc_pred_only_1f_s1%402020-06-18T18%3A07%3A05.771Z&job=test_open_partition_dc_pred_only_s1%402020-06-18T18%3A07%3A15.429Z).

**Comment by ycho** [comment]
AWCY for "--tune psnr" for subset1 at s0 and objective-1-fast at s1 will soon be available at following links:

https://beta.arewecompressedyet.com/?job=master_af7f474fb_1f_s0_psnr_subset1%402020-06-19T20%3A56%3A36.133Z&job=test_open_partition_s0_subset1_psnr%402020-06-19T20%3A58%3A57.092Z

https://beta.arewecompressedyet.com/?job=master_c6bf0bf_s1_low_latency_psnr%402020-06-19T21%3A02%3A47.414Z&job=test_open_partition_rebase3_s1_low_latency_psnr%402020-06-19T21%3A00%3A18.744Z

**Comment by ycho** [comment]
Comparing inter frame (2nd frame), Qindex 172 given to encoder (actual 291)
https://beta.arewecompressedyet.com/runs/master_af7f474fb@2020-05-25T16:43:25.801Z/objective-1-fast/rush_hour_1080p25_60f.y4m-172.ivf

https://beta.arewecompressedyet.com/runs/test_open_partition_btmup_on_boundary_s6@2020-05-27T23:16:14.788Z/objective-1-fast/rush_hour_1080p25_60f.y4m-172.ivf

master vs open-partition

<img width="1750" alt="Screen Shot 2020-06-19 at 3 01 44 PM" src="https://user-images.githubusercontent.com/10123259/85182706-5612a380-b23e-11ea-8ec6-2a3dd00e9744.png">

<img width="1750" alt="Screen Shot 2020-06-19 at 3 01 53 PM" src="https://user-images.githubusercontent.com/10123259/85182713-59a62a80-b23e-11ea-826e-0768c83423ac.png">

**Comment by ycho** [comment]
Regarding whether MV for padded area is used, it seems so.
Please look at bottom-left corner, bi-directional prediction, which has MV (shown right side panel), "Motion Vectors", (0,-70), (0,70), the blue arrow fetches from outside frame, i.e., padde area, I think.

<img width="1750" alt="Screen Shot 2020-06-19 at 3 03 33 PM" src="https://user-images.githubusercontent.com/10123259/85182904-e4872500-b23e-11ea-9721-382c8809e984.png">

**Comment by ycho** [comment]
[test03_146x146.y4m.gz](https://github.com/xiph/rav1e/files/4810064/test03_146x146.y4m.gz)

[test03_248x248.y4m.gz](https://github.com/xiph/rav1e/files/4810066/test03_248x248.y4m.gz)

**Comment by ycho** [comment]
All right, I think we found the clue!
I forgot that I left two lines of code that set deblocking strength as zero,
 left these two lines here : https://github.com/ycho/rav1e/blob/01f57b36019f1e8f0d93a543c5cda233211d6ee4/src/encoder.rs#L3088

Many tines, I turn off as many coding feature possible, including those three in-loop filters when I begin a new video codec task and that can possibly touches or interfere with other coding tools. And this time, the task was large to me and whole my mind and eyes was taken for seeking the reason for coding loss and completely forgot turning back on one of in-loop filters, deblocking filter.

I have not verified it yet, but awcy result comes soon!

**Comment by ycho** [comment]
Okay, my fix was correct and start seeing the coding gains.

[awcy speed 6, no frame reorder](https://beta.arewecompressedyet.com/?job=master_b14cbc73_s6_low_latency_psnr%402020-06-20T21%3A56%3A49.406Z&job=test_open_partition_restore_deblock_s6_low_latency_psnr%402020-06-21T22%3A11%3A18.454Z)

|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -1.5582 | -1.7591 | -1.8580 |  -1.5413 | -1.6655 | -1.5900 |    -1.7712 |

**Comment by ycho** [comment]
[awcy speed 0, subset1](https://beta.arewecompressedyet.com/?job=master_af7f474fb_1f_s0subset1%402020-06-14T15%3A56%3A35.870Z&job=test_open_partition_restore_deblock_s0_subset1%402020-06-21T22%3A39%3A22.203Z)

|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -0.4097 |     N/A | -0.6440 |  -0.3641 | -0.4035 | -0.3906 |    -0.5393 |

**Comment by ycho** [comment]
[awc
[awcy,  speed 1 (all bottom up) with no frame reorder, psnr](https://beta.arewecompressedyet.com/?job=test_open_partition_restore_deblock_s1_low_latency_psnr%402020-06-21T22%3A10%3A52.453Z&job=master_c6bf0bf_s1_low_latency_psnr%402020-06-19T21%3A02%3A47.414Z)

|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -1.6251 | -2.3216 | -2.4041 |  -1.6014 | -1.7150 | -1.6379 |    -2.0424 |

**Comment by ycho** [comment]
If I don't use bottomup search on frame boundary and use the same topdown, there is a big coding loss, because topdown search cannot split and lookahead down more than one level deep when split is required by av1 spec, just largest partition (64x64) is used.

[awcy, default speed 6](https://beta.arewecompressedyet.com/?job=master_b14cbc73_s6%402020-06-21T22%3A15%3A55.738Z&job=test_open_partition_revert_btmup_s6%402020-06-22T00%3A47%3A28.044Z)
This mostly happens with 720p and 360p test sequences.

|   PSNR | PSNR Cb | PSNR Cr | PSNR HVS |   SSIM | MS SSIM | CIEDE 2000 |
|   ---: |    ---: |    ---: |     ---: |   ---: |    ---: |       ---: |
| 2.3520 |  0.6730 |  1.1094 |   1.5370 | 2.3579 |  1.7963 |     1.2295 |

1st frame (key frame), speed 6, master vs open-partition (with not using bottomup on frame boundary for topdown)

<img width="1306" alt="Screen Shot 2020-06-21 at 9 00 42 PM" src="https://user-images.githubusercontent.com/10123259/85247430-93c02980-b402-11ea-82ca-cb1ce2def8c4.png">
<img width="1306" alt="Screen Shot 2020-06-21 at 9 00 47 PM" src="https://user-images.githubusercontent.com/10123259/85247433-96bb1a00-b402-11ea-9122-842a84b53d06.png">

1st frame (i.e, 1st inter frame), speed 6, master vs open-partition (with not using bottomup on frame boundary for topdown)

<img width="1306" alt="Screen Shot 2020-06-21 at 9 01 05 PM" src="https://user-images.githubusercontent.com/10123259/85247440-a0448200-b402-11ea-9702-aedb43b8ee55.png">
<img width="1306" alt="Screen Shot 2020-06-21 at 9 01 10 PM" src="https://user-images.githubusercontent.com/10123259/85247441-a3d80900-b402-11ea-9de7-b429bb62e42b.png">

**Comment by ycho** [comment]
[And most superset awcy result, speed 0](https://beta.arewecompressedyet.com/?job=master_b14cbc73_s0%402020-06-21T22%3A16%3A42.236Z&job=test_open_partition_restore_deblock_s0%402020-06-21T22%3A33%3A46.803Z)

|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -1.8174 | -2.0760 | -2.1534 |  -1.7602 | -1.8175 | -1.7785 |    -2.0376 |

**Comment by ycho** [comment]
speed_bag_640x360_60f.y4m, QIndex 252

master vs open-partition, frame 1 (key frame, intra frame)

<img width="1284" alt="Screen Shot 2020-06-21 at 9 21 19 PM" src="https://user-images.githubusercontent.com/10123259/85248309-4f825880-b405-11ea-8a72-a53a94213a1a.png">
<img width="1284" alt="Screen Shot 2020-06-21 at 9 21 24 PM" src="https://user-images.githubusercontent.com/10123259/85248313-5315df80-b405-11ea-8508-169450e9669a.png">

master vs open-partition, frame 2 (i.e. 1st inter frame)

<img width="1284" alt="Screen Shot 2020-06-21 at 9 21 34 PM" src="https://user-images.githubusercontent.com/10123259/85248349-6cb72700-b405-11ea-8735-e1ec3846710a.png">
<img width="1284" alt="Screen Shot 2020-06-21 at 9 21 39 PM" src="https://user-images.githubusercontent.com/10123259/85248356-70e34480-b405-11ea-80fd-9e66d56700ed.png">

**Comment by ycho** [comment]
As a sanity check,
master vs "open-partition branch with old partition searches"

0.0% change for s6 and +0.0041%  for s2.

[awcy s6](https://beta.arewecompressedyet.com/?job=test_open_partition_deblk_no_btmup3_with_old_partition_s6_low_latency%402020-06-22T18%3A40%3A20.609Z&job=master_b14cbc73_s6_low_latency%402020-06-22T03%3A32%3A43.402Z)

[awcy s6, psnr](https://beta.arewecompressedyet.com/?job=master_b14cbc73_s6_low_latency_psnr%402020-06-20T21%3A56%3A49.406Z&job=test_open_partition_deblk_no_btmup3_with_old_partition_s6_low_latency_psnr%402020-06-22T18%3A41%3A13.044Z)

[awcy s2](https://beta.arewecompressedyet.com/?job=master_b14cbc73_s2_low_latency%402020-06-22T08%3A19%3A05.604Z&job=test_open_partition_deblk_no_btmup3_with_old_partition_s2_low_latency%402020-06-22T18%3A42%3A57.733Z)

[awcy s2, psnr](https://beta.arewecompressedyet.com/?job=master_b14cbc73_s2_low_latency_psnr%402020-06-22T20%3A44%3A47.834Z&job=test_open_partition_deblk_no_btmup3_with_old_partition_s2_low_latency_psnr%402020-06-22T18%3A42%3A36.888Z)

**Comment by ycho** [comment]
Another screenshots of bitstream analyzer, how open-partition contributes to reduce bit rate by using larger partition sizes and thus encoding less amount of MVs in addition to partition info.

speed_bag_640x360_60f.y4m, QIndex 252

master vs open-partition, 5th frame
<img width="1551" alt="Screen Shot 2020-06-23 at 9 50 53 AM" src="https://user-images.githubusercontent.com/10123259/85432164-6aed8080-b537-11ea-9fd2-b546368c1cea.png">
<img width="1551" alt="Screen Shot 2020-06-23 at 9 50 59 AM" src="https://user-images.githubusercontent.com/10123259/85432183-704acb00-b537-11ea-9715-4a6e3998caf1.png">

**Comment by ycho** [comment]
Regarding the regression with topdown partition search, i.e. speed levels 2~10, I think I found the better reason and possible improvement idea for it. Previously, I explained that the reason is  topdown search revised in open-partition branch does not split when it lookaheads that any of its children partition requires mandatary split by av1 spec. This is due to its nature of algorithm and if we ever change it with major revisions so that it can, then we will realize it is doing the task what bottomup search achieves. In a difficult way.
The better reason I've found now is, **topdown for open-partition does not uses non-square partitions at all.**
This means, when topdown is able to split down to smallest possible partition size on left or bottom frame boundaries, even when non-square is best fit in rdo sense, it must use square partitions.
FYI, we stopped using non-square partitions for topdown before. Because using it middle of split leads to worse partition decisions.
We, however, we have had the feature that non-square partitions are only used for terminal cases, i.e. smallest partition sizes allowed.
Now the possible fix is, move the lookahead that checks whether children partition requires mandatary partition (done in rdo_partition_simple( ) ) prior to calling  rdo_partition_decision( ). So that, if mandatary partition requires, then that means PARTITION_SPLIT type does not work, so we add PARTITION_VERT or PARTITION_HORZ (depends on has_cols and has_rows condition) to partition_types parameter that is passed to rdo_partition_decision( ).
